module/documents/actor/actor.mjs

```diff
@@ -627,16 +627,30 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
     // Translate the list of classes into spellcasting progression
     const progression = { slot: 0, pact: 0 };
     const types = {};
+    const preparations = {};
+    let classes = {};
 
     // Grab all classes with spellcasting
-    const classes = this.items.filter(cls => {
-      if ( cls.type !== "class" ) return false;
-      const type = cls.spellcasting.type;
-      if ( !type ) return false;
-      types[type] ??= 0;
-      types[type] += 1;
-      return true;
+    for ( const item of this.items ) {
```

Comment:
A little clumsier than it could be, but I'm trying hard to avoid iterating over the entire item list too many times.

Reply:
We could always do that in `prepareFinalData` in `SpellData`. So it would do something like:

```javascript
const config = CONFIG.DND5E.spellPreparationModes[this.preparation.mode];
if ( this.parent.isOwned && this.sourceClass && config?.prepares && ((this.preparation.mode === "always") || prepared ) {
  const sourceClass = this.parent.actor.spellcastingClasses[this.sourceClass];
  if ( sourceClass ) sourceClass.system.spellcasting.preparation.value += 1;
}
```

We are already calling this for spells so there is no additional loops performed, and the system ensures that `prepareFinalData` is always called once.

Reply:
`actor.itemTypes` is cached now, so that could help.